### PR TITLE
CSP-482-GET Employer Member

### PR DESCRIPTION
**Pull request by ravihejmadi**
New Query,Result, UnitTests
Modified Repo, Controllers, Tests
Implemented New ValidatedResponse FluentValidation Middleware
New Branch with new PR with Code Review fixes from https://github.com/SkillsFundingAgency/das-aan-hub-api/pull/38

image

Additional change:
Removed nullability check of `requestedByUserId` as application will not allow null, empty, or incorrect values to enter controller method anyway.

Example bad request responses:

```
{
    "type": "https://tools.ietf.org/html/rfc7231#section-6.5.1",
    "title": "One or more validation errors occurred.",
    "status": 400,
    "traceId": "00-37d3bec6256270d8c773e332584c1d27-990913ed73300666-00",
    "errors": {
        "x-requested-by-userid": [
            "The value '' is invalid."
        ]
    }
}
```

```
{
    "type": "https://tools.ietf.org/html/rfc7231#section-6.5.1",
    "title": "One or more validation errors occurred.",
    "status": 400,
    "traceId": "00-37d3bec6256270d8c773e332584c1d27-990913ed73300666-00",
    "errors": {
        "x-requested-by-userid": [
            "The value 'abcd123' is invalid."
        ]
    }
}
```

### CODE REVIEW

File: src/SFA.DAS.AANHub.Application/Employers/Queries/GetEmployerMemberResult.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        public string Name { get; set; } = null!;
+        public string Organisation { get; set; } = null!;
+        public static implicit operator GetEmployerMemberResult?(Employer? employer)
+        {
+            if (employer == null)
+                return null;
+
+            return new GetEmployerMemberResult
+            {
+                Email = employer.Email!,
```
Comment (LiamWright) [inline src/SFA.DAS.AANHub.Application/Employers/Queries/GetEmployerMemberResult.cs:18]: I don't think you need the nullable flag here if the properties have a default value, and you've already checked to see if Employer is null.

**Review by LiamWright — COMMENTED** [review]
Just a small comment

File: src/SFA.DAS.AANHub.Api.UnitTests/Controllers/EmployersControllerTests.cs [hunk 2]
```diff
@@ -71,5 +71,101 @@ public EmployersControllerTests()
...
+            var response = await _controller.GetEmployer(userRef);
+            response.Should().NotBeNull();
+
+            var result = response as OkObjectResult;
+            Assert.AreEqual(StatusCodes.Status200OK, result!.StatusCode);
+
+            var queryResult = result.Value as GetEmployerMemberResult;
+            queryResult.Should().BeEquivalentTo(handlerResult.Result);
+
+            _mediator.Verify(m => m.Send(It.IsAny<GetEmployerMemberQuery>(), It.IsAny<CancellationToken>()));
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Api.UnitTests/Controllers/EmployersControllerTests.cs:92]: Could verify that the passed in UserRef is set in the query by using ~ `It.Is<GetEmployerMemberQuery>(q => q.UserRef == userRef)`

File: src/SFA.DAS.AANHub.Api.UnitTests/Controllers/EmployersControllerTests.cs [hunk 3]
```diff
@@ -71,5 +71,101 @@ public EmployersControllerTests()
...
+            var userRef = Guid.NewGuid();
+            _mediator.Setup(m => m.Send(It.Is<GetEmployerMemberQuery>(q => q.UserRef == userRef),
+                    It.IsAny<CancellationToken>()))
+                .ReturnsAsync(response);
+
+            var getResult = await _controller.GetEmployer(userRef);
+
+            var result = getResult as OkObjectResult;
+            result.Should().NotBeNull();
+            Assert.AreEqual(StatusCodes.Status200OK, result!.StatusCode);
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Api.UnitTests/Controllers/EmployersControllerTests.cs:137]: Redundant verification, status code 200 is implicit with OkObjectResult. I would rather verify the result.Value has expected outcome.

File: src/SFA.DAS.AANHub.Application/Employers/Commands/CreateEmployerMemberCommandValidator.cs [hunk 4]
```diff
@@ -7,7 +7,7 @@ namespace SFA.DAS.AANHub.Application.Employers.Commands
 {
     public class CreateEmployerMemberCommandValidator : AbstractValidator<CreateEmployerMemberCommand>
     {
-        private const string EmployerPairAlreadyCreatedErrorMessage = "UserId and AccountId pair already exist";
+        private const string EmployerPairAlreadyCreatedErrorMessage = "UserRef already exists";
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Application/Employers/Commands/CreateEmployerMemberCommandValidator.cs:10]: Variable name can be updated to not have pair any more.

File: src/SFA.DAS.AANHub.Application/Employers/Commands/CreateEmployerMemberCommandValidator.cs [hunk 5]
```diff
@@ -18,27 +18,22 @@ public class CreateEmployerMemberCommandValidator : AbstractValidator<CreateEmpl
...
-
-            RuleFor(c => c.UserId)
-                .NotEmpty()
-                .MustAsync(async (command, _, _) => await CheckAccountIdAndUserIdPair(command.AccountId, command.UserId))
-                .WithMessage(EmployerPairAlreadyCreatedErrorMessage);
+                .NotEmpty();
         }
 
-        private async Task<bool> CheckAccountIdAndUserIdPair(long accountId, long userId)
+        private async Task<bool> CheckUserRefExists(Guid userRef)
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Application/Employers/Commands/CreateEmployerMemberCommandValidator.cs:34]: `CheckUserRefExists` = true could mean that it exists, however true means it does not. So I suggest the method be renamed in line with `IsNewUser`.

File: src/SFA.DAS.AANHub.Api/Controllers/EmployersController.cs [hunk 6]
```diff
@@ -31,12 +32,12 @@ public EmployersController(ILogger<EmployersController> logger, IMediator mediat
         [Produces("application/json")]
         [ProducesResponseType(typeof(string), StatusCodes.Status400BadRequest)]
         [ProducesResponseType(typeof(string), StatusCodes.Status201Created)]
-        public async Task<IActionResult> CreateEmployer([FromHeader(Name = Constants.PostRequestHeaders.RequestedByUserHeader)] [Required] Guid? userId,
+        public async Task<IActionResult> CreateEmployer([FromHeader(Name = Constants.PostRequestHeaders.RequestedByUserHeader)] [Required] Guid userId,
             CreateEmployerModel request)
         {
             _logger.LogInformation("AAN Hub API: Received command to add employer by accountId: {accountId} and UserId: {userId}",
                 request.AccountId,
-                request.UserId);
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Api/Controllers/EmployersController.cs:39]: It is worth to include the requesting user id in the log

File: src/SFA.DAS.AANHub.Application/Employers/Queries/GetEmployerMemberResult.cs [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+namespace SFA.DAS.AANHub.Application.Employers.Queries
+{
+    public class GetEmployerMemberResult
+    {
+        public Guid MemberId { get; set; }
+        public string Email { get; set; } = null!;
+        public string Name { get; set; } = null!;
+        public string Organisation { get; set; } = null!;
+
+        public static implicit operator GetEmployerMemberResult?(Employer? employer)
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Application/Employers/Queries/GetEmployerMemberResult.cs:12]: I am not liking this null checks, if the argument is made not nullable then we can avoid this null checks.

File: src/SFA.DAS.AANHub.Application/SFA.DAS.AANHub.Application.csproj [hunk 8]
```diff
@@ -21,6 +21,7 @@
   <ItemGroup>
     <Folder Include="Apprentices\Commands\" />
     <Folder Include="Common\Validators\RequestedByMemberId\" />
+    <Folder Include="Employers\Queries\" />
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Application/SFA.DAS.AANHub.Application.csproj:24]: This is definitely not needed, perhaps it got added when the folder was created with no files.

File: src/SFA.DAS.AANHub.Data/Repositories/EmployersReadRepository.cs [hunk 9]
```diff
@@ -12,9 +12,9 @@ internal class EmployersReadRepository : IEmployersReadRepository
 
         public EmployersReadRepository(AanDataContext aanDataContext) => _aanDataContext = aanDataContext;
 
-        public async Task<Employer?> GetEmployerByAccountIdAndUserId(long accountId, long userId) => await _aanDataContext
+        public async Task<Employer?> GetEmployerByUserRef(Guid userRef) => await _aanDataContext
             .Employers
-            .AsNoTracking().Where(m => m.AccountId == accountId && m.UserId == userId)
+            .AsNoTracking().Where(m => m.UserRef == userRef)
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Data/Repositories/EmployersReadRepository.cs:17]: Where on new line to keep it consistent

File: src/SFA.DAS.AANHub.Database/Tables/Employer.sql [hunk 10]
```diff
@@ -2,10 +2,10 @@
...
-    [Organisation] NVARCHAR(250) NULL, 
-    [Email] NVARCHAR(256) NULL, 
-    [Name] NVARCHAR(200) NULL, 
-    [LastUpdated] DATETIME NULL
+    [UserRef] UNIQUEIDENTIFIER NOT NULL, 
+    [Organisation] NVARCHAR(250) NOT NULL, 
+    [Email] NVARCHAR(256) NOT NULL, 
+    [Name] NVARCHAR(200) NOT NULL, 
+    [LastUpdated] DATETIME NOT NULL
     CONSTRAINT [FK_Employer_Member] FOREIGN KEY ([MemberId]) REFERENCES [Member]([Id])
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Database/Tables/Employer.sql:10]: Need to add a unique constraint `CONSTRAINT [UK_Employer] UNIQUE ([UserRef])`

File: src/SFA.DAS.AANHub.Application.UnitTests/SFA.DAS.AANHub.Application.UnitTests.csproj [hunk 11]
```diff
@@ -38,6 +38,7 @@
 
   <ItemGroup>
     <Folder Include="Apprentices\Commands\" />
+    <Folder Include="Employers\Queries\" />
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Application.UnitTests/SFA.DAS.AANHub.Application.UnitTests.csproj:41]: Not sure why these Folder attributes are getting into project file, if there are files in the folder then this should be removed.

**Review by DevenKShah — COMMENTED** [review]
Nice work removing the nullable on Requesting user id. 👍 
I have got few comments on here but all of them are of minor suggestive nature except that you missed unique constraint on the table.

File: src/SFA.DAS.AANHub.Api.UnitTests/Controllers/EmployersControllerTests.cs (hunk 2 above)
Comment (LiamWright) [reply src/SFA.DAS.AANHub.Api.UnitTests/Controllers/EmployersControllerTests.cs]: Test has been updated with suggestion

File: src/SFA.DAS.AANHub.Api.UnitTests/Controllers/EmployersControllerTests.cs (hunk 3 above)
Comment (LiamWright) [reply src/SFA.DAS.AANHub.Api.UnitTests/Controllers/EmployersControllerTests.cs]: Test now compares result, and mocked response.

File: src/SFA.DAS.AANHub.Application/Employers/Commands/CreateEmployerMemberCommandValidator.cs (hunk 4 above)
Comment (LiamWright) [reply src/SFA.DAS.AANHub.Application/Employers/Commands/CreateEmployerMemberCommandValidator.cs]: Missed that! Updated.

File: src/SFA.DAS.AANHub.Application/Employers/Commands/CreateEmployerMemberCommandValidator.cs (hunk 5 above)
Comment (LiamWright) [reply src/SFA.DAS.AANHub.Application/Employers/Commands/CreateEmployerMemberCommandValidator.cs]: Method name updated

File: src/SFA.DAS.AANHub.Api/Controllers/EmployersController.cs (hunk 6 above)
Comment (LiamWright) [reply src/SFA.DAS.AANHub.Api/Controllers/EmployersController.cs:39]: Added requesting user id into the log

File: src/SFA.DAS.AANHub.Application/Employers/Queries/GetEmployerMemberResult.cs (hunk 7 above)
Comment (LiamWright) [reply src/SFA.DAS.AANHub.Application/Employers/Queries/GetEmployerMemberResult.cs:12]: So, I think the reason we need to do this is that ValidationBehaviour doesn't accept null result types yet.
` return new ValidatedResponse<GetEmployerMemberResult?>(employer);`

So this wouldn't work. I agree with the nullability checks though, they need to be removed.

File: src/SFA.DAS.AANHub.Data/Repositories/EmployersReadRepository.cs (hunk 9 above)
Comment (LiamWright) [reply src/SFA.DAS.AANHub.Data/Repositories/EmployersReadRepository.cs]: Done

File: src/SFA.DAS.AANHub.Database/Tables/Employer.sql (hunk 10 above)
Comment (LiamWright) [reply src/SFA.DAS.AANHub.Database/Tables/Employer.sql:10]: Added

**Review by DevenKShah — APPROVED** [review]

**Pull request merged**